### PR TITLE
Refactor codebase to use modern JS well

### PR DESCRIPTION
**Pull request by n8xm**
1. [ES6 modules](https://hacks.mozilla.org/2015/08/es6-in-depth-modules/) wherever possible.
2. [Doug Crockford's](https://en.wikipedia.org/wiki/Douglas_Crockford)'s "[Class Free OOP](https://gist.github.com/mpj/17d8d73275bca303e8d2)" is used instead of classes.
    1. Classes and `new` in JS were always half baked [1][2]
    2. Allows us to make variables and methods private (impossible with traditional classes in JS). 
    3. Makes `this` unnecessary. (`this` in JS breaks encapsulation and also makes code more verbose.)
3. Now, a `ControlsManager` processes inputs from the sliders and calls handlers on the `BRDFViewer` and `ModelViewer`. Previously, the `BRDFViewer` handled all input and modified `ModelViewer`'s state. **The `BRDFViewer` and `ModelViewer` should not modify each others' state.**  In theory, they shouldn't be aware of each other at all, because despite having common input parameters they are independent views.
4. Variables are now delcared at the top of their scope (this is [suggested by w3schools](https://www.w3schools.com/js/js_hoisting.asp) and [1] due to [hoisting](https://www.w3schools.com/js/js_hoisting.asp)). 
5. I am beginning to use `let` instead of `var` when I want block scope, because unlike in other languages, [JS variables declared with `var` do not have block scope](https://www.sitepoint.com/joys-block-scoping-es6/), and this can lead to counterintuitive behavior. `let`, on the other hand, [declares a block scope variable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let).
6. I am trying to use `undefined` instead of `null` (ideally, getting rid of `null` so that we only have one value that means "empty"), and I am also attempting to [avoid falsiness.](https://developer.mozilla.org/en-US/docs/Glossary/Falsy) [2] suggests that we do this. 

[1] Crockford, Doug. _Javascript: The Good Parts._ 2008.
[2] Crockford, Doug. _Javascript: The Better Parts._ code::dive conference. 2017. <https://youtu.be/DxnYQRuLX7Q>

### CODE REVIEW

**Comment by n8vm** [comment]
I’m a bit leery about using Crockford Classless. 

This small Reddit post pretty much matches my sentiment: https://www.google.com/amp/s/amp.reddit.com/r/javascript/comments/44dta3/classless_javascript_by_douglas_crockford/


Especially where this project is somewhat small, I don’t think it’s a good idea to enforce unstandardized js practices.

File: Scripts/ModelViewport.js [hunk 1]
```diff
@@ -7,132 +7,58 @@
...
-        //var fragmentShader = this.getShader(gl, "shader-fs");
-        //var vertexShader = this.getShader(gl, "shader-vs");
-
-        //this.shaderProgram = gl.createProgram();
-        //gl.attachShader(this.shaderProgram, vertexShader);
-        //gl.attachShader(this.shaderProgram, fragmentShader);
-        //gl.linkProgram(this.shaderProgram);
-
-        //if (!gl.getProgramParameter(this.shaderProgram, gl.LINK_STATUS)){
+export default function ModelViewport(spec) {
```
Comment (n8vm) [inline Scripts/ModelViewport.js]: I personally am not a fan of this. ModelViewport should be an object, not a function. Code should be self commenting, and this feels misleading. 

I understand that javascript classes are a bit lacking though. Like, having to use "this" all the time sucks.

Could you put a block comment above functions that we should be pretending are classes/objects?

Will this effect callbacks? For example, "function() {}" vs "() => {}" syntax might suddenly change what the keyword "this" refers to. If I wanted to write a callback, I sometimes pass the keyword "this" as a parameter. Now I don't think I can do that...

File: Scripts/main.js [hunk 2]
```diff
@@ -1,15 +1,30 @@
...
-  requestAnimationFrame(render);
-}
-
-/* Requires BRDFViewport.js */
-document.addEventListener('DOMContentLoaded', function () {
-  brdfViewport = new BRDFViewport("brdf-canvas", 512, 512);
-  modelViewport = new ModelViewport("model-canvas", 512, 512);
-  requestAnimationFrame(render);
-});
+"use strict";
```
Comment (n8vm) [inline Scripts/main.js]: The lines below break on my version of chrome:
I get:
"Failed to load module script: The server responded with a non-JavaScript MIME type of "text/plain". Strict MIME type checking is enforced for module scripts per HTML spec."

File: Scripts/ControlsManager.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
+"use strict";
+
+//requires d3.js 
+
+export default function ControlsManager(){
+  const starting_theta = 45;
+  const starting_phi = 0;
+
+  let
```
Comment (n8vm) [inline Scripts/ControlsManager.js]: What is the purpose of "let" here?

File: Scripts/ControlsManager.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        });
+      };
+    };
+
+  //************* Start "constructor" **************
+  setupUI();
+  setupUICallbacks();
+  //************* End "constructor" **************
+
+  return Object.freeze({
```
Comment (n8vm) [inline Scripts/ControlsManager.js]: what is the purpose of this?

File: index.html [hunk 5]
```diff
@@ -18,37 +19,8 @@
...
-   -    void main() {
-   -        vColor = color;
-   -        gl_Position = u_p * u_v * u_m * model_position;
-   -        //world_normal = u_m * model_normal;
-   -        world_normal = model_normal; //DEBUG ONLY.
-   -    }
-   -  </script>
-   -->
-
   <script type="vertex" id="lobe.vert">#version 300 es
```
Comment (n8vm) [inline index.html]: Can we get these shaders moved out of index.html? Maybe load them in using AJAX instead?

File: known_issues.txt [hunk 6]
```diff
@@ -1,2 +1,4 @@
-1) We should probably change gl.DYNAMIC_DRAW calls to gl.STATIC_DRAW calls when we are done doing the lobe on GPU.
-2) There is some sort of... seam when we set the color using polar_to_color. It's definitely not a mach band... does it appear near phi=0?
+1) There is some sort of... seam when we set the color using polar_to_color. It's definitely not a mach band... does it appear near phi=0?
+2) Currently we put our shaders inside index.html... ideally, especially once our shaders get larger, we will want to put them in their own standalone files. But in order to do that, it looks like we'll have to make an asynchronous XMLHttpRequest (see network-wranglers.js).
```
Comment (n8vm) [inline known_issues.txt]: Yup. jquery can make AJAX a little cleaner. 
https://www.w3schools.com/jquery/ajax_ajax.asp

File: known_issues.txt [hunk 7]
```diff
@@ -1,2 +1,4 @@
...
+3) Currently, our GUI controls are bolted on to the BRDFViewer... I believe that the "correct" way to do it is to have a third object store the GUI state (i.e. the model). The renderes function as views. Finally, a controller would modify the GUI state. For our application we can probably merge view and controller.
```
Comment (n8vm) [inline known_issues.txt]: Yup. I've usually implemented dynamic guis with a controller and either d3, angular, or jquery.

File: known_issues.txt [hunk 8]
```diff
@@ -1,2 +1,4 @@
...
+4) Currently Firefox doesn't fully support ES6 modules. Therefore, our app works if dom.moduleScripts.enabled is set to "true" in about.config.
```
Comment (n8vm) [inline known_issues.txt]: I can't seem to get the newer syntax to run in either chrome or firefox.

Firefox gives "Loading failed for the <script> with source “http://localhost:8000/Scripts/main.js”."

File: Scripts/network-wranglers.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+export function loadTextFile(path){
```
Comment (n8vm) [inline Scripts/network-wranglers.js]: Is there a reason to use this over jquery AJAX?

File: Scripts/main.js [hunk 10]
```diff
@@ -1,15 +1,30 @@
...
+
+import BRDFViewport from "./BRDFViewport.js";
+import ModelViewport from "./ModelViewport.js";
+import ControlsManager from "./ControlsManager.js";
+
+//Wrap everything in a function so that we do not pollute
+//the global namespace.
```
Comment (n8vm) [inline Scripts/main.js]: I don't mind polluting the global namespace a bit. I actually prefer having access to these objects from the javascript console, and would prefer they be global.

Just want to avoid polluting when collisions will inevitably happen, like when declaring a variable "gl" as a global.

**Comment by n8xm** [comment]
Oops. @n8vm @DQLin . This is looking like it was a bad idea.

File: Scripts/main.js (hunk 2 above)
Comment (n8xm) [reply Scripts/main.js]: The most likely reason why things would break is ES6 modules, and a [the first result on a quick Google search seems to confirm this](https://stackoverflow.com/questions/47403478/es6-modules-in-local-files-the-server-responded-with-a-non-javascript-mime-type). In Chrome they should work but if you load the files locally the same-origin policy causes @n8vm 's error message.

I may want to leave ES6 modules out because even though they should work with the latest Chrome, they still don't work out of the box in the latest Firefox (and possibly other browsers)?

Comment (n8vm) [reply Scripts/main.js]: Hmm, I’m thinking it might have something to do with how i’m hosting my local server. 

How have you been serving the files on your machine? I’ve been using python -m http.server.

**Comment by n8xm** [comment]
@n8vm On Crockford classes: Since they are unpleasant to swallow I'll go back to what we had before. I defer to you because you are almost certainly better at web programming than I am. 

`<nerdrage> `

That being said, I don't think they are quuuiiiitttteee as bad as [people on Reddit](https://www.reddit.com/r/javascript/comments/44dta3/classless_javascript_by_douglas_crockford/) think they are. 

None of the commenters seem to care about keeping variables private. I consider that to be a pretty significant advantage to Crockford's approach. 

The first commenter tells us that Crockford classes are unnecessary because we can use composition instead of inhertiance. However, if you don't use pseudoclassical inheritance, then the first commenter's way of doing this is also unnecessary. To do composition or prototypal inheritance, you don't need the _class_ keyword or the _new_ operator at all. In fact, the _class_ keyword is a new feature in ES6, and its purpose is mostly syntactic sugar to make JS look more like other languages. 

The second commenter seems to be saying that using options objects without destructuring is unnecessary. If so, they can easily be dropped from the Crockford classes. 

The third commenter says that Crockford classes are convoluted and contorted just for the sake of being different. I disagree because they are a slight variation on [revealing module pattern (RMP)](http://jargon.js.org/_glossary/REVEALING_MODULE_PATTERN.md), which is a well-known use case for closures in JS. Both closures and the RMP are well-known and have been around for years. (E.g. here is a [blog post on RMP from over 10 years ago](https://christianheilmann.com/2007/08/22/again-with-the-module-pattern-reveal-something-to-the-world/).) I would argue that Crockford's pattern only appears convoluted to someone who is unfamiliar with those two concepts. 

`<\nerdrage> `

Okay, nerd rage over. Since it's a small project I'll just go back to "vanilla" classes.

File: Scripts/main.js (hunk 2 above)
Comment (n8xm) [reply Scripts/main.js]: I am also using `python -m http.server`.

File: Scripts/ControlsManager.js (hunk 3 above)
Comment (n8xm) [reply Scripts/ControlsManager.js]: Here it does the same thing as "var", because in this case block scope and function scope are the same. This is my bad because I should have been consistent with this (in the other two classes they are "var").

File: Scripts/ControlsManager.js (hunk 4 above)
Comment (n8xm) [reply Scripts/ControlsManager.js]: The freeze prevents us from adding new methods / properties onto the object after we've created it. There is a performance argument for doing this (modifying objects at runtime can kill performance, see [WebGL Insights Ch. 4](http://webglinsights.github.io/downloads/WebGL-Insights-Chapter-4.pdf)), and a style argument for doing this (behaves more like a class in other languages, maybe code is less clean if we are adding properties in several different places).

File: known_issues.txt (hunk 6 above)
Comment (n8xm) [reply known_issues.txt]: :+1: The next thing I'll do is use jquery to load the shaders.

File: Scripts/network-wranglers.js (hunk 9 above)
Comment (n8xm) [reply Scripts/network-wranglers.js]: Not especially. I was scared of using a powerful library that does lots of things, just to do one thing. E.g. in C++ I've been warned against using Boost unless totally necessary because "before you know it everything will depend on Boost and it will become impossible to get rid of". But lacking experience in JS / jquery, I'll defer to you.

**Comment by n8xm** [comment]
1) Classless syntax is now better-documented. 
2) Shaders have been separated from index.html. 
 i) Model view shader now [conforms to ES 3.00 spec](https://developer.apple.com/library/content/documentation/3DDrawing/Conceptual/OpenGLES_ProgrammingGuide/AdoptingOpenGLES3/AdoptingOpenGLES3.html). 
3) Small refactoring of variable declarations. Now all variables are declared at the top of their scope. Also, 
 `let` is used instead of `var` unless we have a situation where we need `var`'s looser scope. (`let` has block scope, whereas `var` ignores blocks and has function scope.) 
4) Small refactoring of for loops - if we use `Object.keys(object).forEach`, we don't have to do a `hasOwnProperty()` check.

File: Scripts/ModelViewport.js [hunk 11]
```diff
@@ -281,8 +269,39 @@ export default function ModelViewport(spec) {
   canvas.width = width;
   canvas.height = height;
   setupWebGL2();
-  initShaders();
-  loadModels();
+
+  const shdrDir = "Shaders/"; //FIXME: duplicated code from BRDFViewport
```
Comment (n8vm) [inline Scripts/ModelViewport.js]: Little thing, this should be ./Shaders/, so that if the project is hosted in a subdirectory, these will be located relative to index.html.

**Comment by DQLin** [comment]
The refactor seems ok.
